tests/qactivation_test.py

```diff
@@ -517,6 +519,235 @@ def test_quantized_bits(bits, integer, symmetric, keep_negative, test_values,
   result = f([test_values])[0]
   assert_allclose(result, expected_values, rtol=rtol)
 
+class TestQuantizedLinear:
+  """Tests for quantized_linear"""
+
+  def test_sign_function(self):
+    "Test to make sure that sign function is working properly"
+
+    quantizer = quantized_linear(bits=1, keep_negative=True)
+    x = tf.constant([-1.0, 0.0, 1.0, 2.0, 3.0])
+
+    res = quantizer(x)
+    expected_res = quantizer.max() * tf.constant([-1.0, 1.0, 1.0, 1.0, 1.0])
+
+    assert tf.reduce_all(tf.equal(res, expected_res))
+
+  @pytest.mark.parametrize('shape', [(1,), (2, 3), (2, 2, 4)])
+  def test_scale_shape(self, shape):
+    """Test to make sure that scale is the right shape for auto-alphas"""
+
+    auto_quantizer = quantized_linear(alpha='auto')
+    auto_po2_quantizer = quantized_linear(alpha='auto_po2')
+    x = tf.ones(shape)
+    auto_quantizer(x)
+    auto_po2_quantizer(x)
+
+    if len(shape) == 1:
+      expected_shape = (1,)
+    else:
+      ones = [1 for _ in range(len(shape) - 1)]
+      expected_shape = tuple(ones) + shape[-1:]
+
+    assert auto_quantizer.scale.shape == expected_shape
+    assert auto_po2_quantizer.scale.shape == expected_shape
+
+  @pytest.mark.parametrize('alpha', [None, 2.0])
+  @pytest.mark.parametrize('symmetric,keep_negative', 
+                          [(True, True), (False, True), (False, False)])
+  @pytest.mark.parametrize('bits', [1, 8])
+  def test_gradients(self, bits, symmetric, keep_negative, alpha):
+
+    quantizer = quantized_linear(bits=bits, symmetric=symmetric,
+                                 keep_negative=keep_negative, alpha=alpha)
+    x = tf.Variable([-1.0, 0.0, 1.0, 2.0, 3.0])
+
+    with tf.GradientTape() as tape:
+      res = quantizer(x)
+
+    grad = tape.gradient(res, x)
+    expected_grad = ((x >= quantizer.min()) & (x <= quantizer.max()))
```

Comment:
Consider add tests on the gradient values themselves.

Reply:
Done